File: 0224-basic-calculator/0224-basic-calculator.java
class Solution {
    public int calculate(String s) {
        int result = 0;
        int sign = 1;
        int num = 0;
        Stack<Integer> stack = new Stack<>();

        for (char c : s.toCharArray()) {
            if (Character.isDigit(c)) {
                num = (num * 10) + (c - '0');
            } else if (c == '+') {
                result += num * sign;
                num = 0;
                sign = 1;
            } else if (c == '-') {
                result += num * sign;
                num = 0;
                sign = -1;
            } else if (c == '(') {
                stack.push(result);
                stack.push(sign);
                result = 0;
                sign = 1;
            } else if (c == ')') {
                result += num * sign;
                num = 0;
                sign = stack.pop();
                int prevResult = stack.pop();
                result = prevResult + sign * result;
            }
        }

        result += num * sign;

        return result;
    }
}

File: 0224-basic-calculator/0224-basic-calculator.kt
class Solution {
    fun calculate(s: String): Int {
        var result = 0
        var sign = 1
        var num = 0
        val stack = ArrayDeque<Int>()

        for (c in s) {
            when (c) {
                in '0'..'9' -> {
                    num = num * 10 + (c - '0')
                }
                '+' -> {
                    result += num * sign
                    num = 0
                    sign = 1
                }
                '-' -> {
                    result += num * sign
                    num = 0
                    sign = -1
                }
                '(' -> {
                    stack.addLast(result)
                    stack.addLast(sign)
                    result = 0
                    sign = 1
                }
                ')' -> {
                    result += num * sign
                    num = 0
                    val prevSign = stack.removeLast()
                    val prevResult = stack.removeLast()
                    result = prevResult + prevSign * result
                }
            }
        }

        result += num * sign

        return result
    }
}